### ISSUE TITLE
Update to Temurin-17.0.5+8 breaks application

### ISSUE THREAD

**Issue by veita** [adoptium/adoptium-support#634]
The application uses signed JARs which causes the `java.util.jar.JarVerifier` to call itself during the class initialization process which leads to a `NullPointerException`. As a result the JVM refuses to start.

The problem disappears when SHA-1 is removed from `jdk.jar.disabledAlgorithms` in `conf/security/java.security`. However, as far as I can tell none of the involved JARs was signed using SHA-1.

The bug is probably related to https://bugs.openjdk.org/browse/JDK-8269039

```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.base/sun.security.util.DisabledAlgorithmConstraints.jarConstraints(DisabledAlgorithmConstraints.java:112)
	at java.base/sun.security.pkcs.SignerInfo.<clinit>(SignerInfo.java:61)
	at java.base/sun.security.pkcs.PKCS7.parseSignedData(PKCS7.java:380)
	at java.base/sun.security.pkcs.PKCS7.parse(PKCS7.java:174)
	at java.base/sun.security.pkcs.PKCS7.parse(PKCS7.java:142)
	at java.base/sun.security.pkcs.PKCS7.<init>(PKCS7.java:124)
	at java.base/sun.security.util.SignatureFileVerifier.<init>(SignatureFileVerifier.java:118)
	at java.base/java.util.jar.JarVerifier.processEntry(JarVerifier.java:308)
	at java.base/java.util.jar.JarVerifier.update(JarVerifier.java:239)
	at java.base/java.util.jar.JarFile.initializeVerifier(JarFile.java:762)
	at java.base/java.util.jar.JarFile.getInputStream(JarFile.java:845)
	at java.base/jdk.internal.util.jar.JarIndex.getJarIndex(JarIndex.java:121)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$1.run(URLClassPath.java:770)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$1.run(URLClassPath.java:762)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader.ensureOpen(URLClassPath.java:761)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader.<init>(URLClassPath.java:735)
	at java.base/jdk.internal.loader.URLClassPath$3.run(URLClassPath.java:499)
	at java.base/jdk.internal.loader.URLClassPath$3.run(URLClassPath.java:482)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:481)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:449)
	at java.base/jdk.internal.loader.URLClassPath$1.next(URLClassPath.java:345)
	at java.base/jdk.internal.loader.URLClassPath$1.hasMoreElements(URLClassPath.java:356)
	at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasNext(BuiltinClassLoader.java:408)
	at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasMoreElements(BuiltinClassLoader.java:416)
	at java.base/java.lang.CompoundEnumeration.next(ClassLoader.java:2725)
	at java.base/java.lang.CompoundEnumeration.hasMoreElements(ClassLoader.java:2734)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1210)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1228)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at java.base/jdk.internal.logger.BootstrapLogger$DetectBackend$1.run(BootstrapLogger.java:897)
	at java.base/jdk.internal.logger.BootstrapLogger$DetectBackend$1.run(BootstrapLogger.java:891)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at java.base/jdk.internal.logger.BootstrapLogger$DetectBackend.<clinit>(BootstrapLogger.java:891)
	at java.base/jdk.internal.logger.BootstrapLogger.useLazyLoggers(BootstrapLogger.java:939)
	at java.base/jdk.internal.logger.LazyLoggers.getLazyLogger(LazyLoggers.java:441)
	at java.base/jdk.internal.logger.LazyLoggers.getLogger(LazyLoggers.java:416)
	at java.base/java.lang.System.getLogger(System.java:1769)
	at java.base/jdk.internal.event.EventHelper.isLoggingSecurity(EventHelper.java:148)
	at java.base/sun.security.provider.X509Factory.commitEvent(X509Factory.java:773)
	at java.base/sun.security.provider.X509Factory.engineGenerateCertificate(X509Factory.java:108)
	at java.base/java.security.cert.CertificateFactory.generateCertificate(CertificateFactory.java:355)
	at java.base/sun.security.pkcs.PKCS7.parseSignedData(PKCS7.java:317)
	at java.base/sun.security.pkcs.PKCS7.parse(PKCS7.java:174)
	at java.base/sun.security.pkcs.PKCS7.parse(PKCS7.java:142)
	at java.base/sun.security.pkcs.PKCS7.<init>(PKCS7.java:124)
	at java.base/sun.security.util.SignatureFileVerifier.<init>(SignatureFileVerifier.java:118)
	at java.base/java.util.jar.JarVerifier.processEntry(JarVerifier.java:308)
	at java.base/java.util.jar.JarVerifier.update(JarVerifier.java:239)
	at java.base/java.util.jar.JarFile.initializeVerifier(JarFile.java:762)
	at java.base/java.util.jar.JarFile.getInputStream(JarFile.java:845)
	at java.base/jdk.internal.util.jar.JarIndex.getJarIndex(JarIndex.java:121)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$1.run(URLClassPath.java:770)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$1.run(URLClassPath.java:762)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader.ensureOpen(URLClassPath.java:761)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader.<init>(URLClassPath.java:735)
	at java.base/jdk.internal.loader.URLClassPath$3.run(URLClassPath.java:499)
	at java.base/jdk.internal.loader.URLClassPath$3.run(URLClassPath.java:482)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:481)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:449)
	at java.base/jdk.internal.loader.URLClassPath$1.next(URLClassPath.java:345)
	at java.base/jdk.internal.loader.URLClassPath$1.hasMoreElements(URLClassPath.java:356)
	at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasNext(BuiltinClassLoader.java:408)
	at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasMoreElements(BuiltinClassLoader.java:416)
	at java.base/java.lang.CompoundEnumeration.next(ClassLoader.java:2725)
	at java.base/java.lang.CompoundEnumeration.hasMoreElements(ClassLoader.java:2734)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1210)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1228)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at org.slf4j.LoggerFactory.findServiceProviders(LoggerFactory.java:104)
	at org.slf4j.LoggerFactory.bind(LoggerFactory.java:147)
	at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:139)
	at org.slf4j.LoggerFactory.getProvider(LoggerFactory.java:422)
	at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:408)
	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:357)
	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:383)
	at de.uplanet.lucy.server.portalserver.PortalService.<clinit>(PortalService.java:84)
Caused by: java.lang.NullPointerException: Cannot invoke "sun.security.util.DisabledAlgorithmConstraints.permits(String, java.security.AlgorithmParameters, sun.security.util.ConstraintsParameters, boolean)" because "sun.security.pkcs.SignerInfo.JAR_DISABLED_CHECK" is null
	at java.base/sun.security.pkcs.SignerInfo.verifyAlgorithms(SignerInfo.java:761)
	at java.base/sun.security.util.SignatureFileVerifier.processImpl(SignatureFileVerifier.java:324)
	at java.base/sun.security.util.SignatureFileVerifier.process(SignatureFileVerifier.java:282)
	at java.base/java.util.jar.JarVerifier.processEntry(JarVerifier.java:327)
	at java.base/java.util.jar.JarVerifier.update(JarVerifier.java:239)
	at java.base/java.util.jar.JarFile.initializeVerifier(JarFile.java:762)
	at java.base/java.util.jar.JarFile.getInputStream(JarFile.java:845)
	at java.base/jdk.internal.util.jar.JarIndex.getJarIndex(JarIndex.java:121)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$1.run(URLClassPath.java:770)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$1.run(URLClassPath.java:762)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader.ensureOpen(URLClassPath.java:761)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader.<init>(URLClassPath.java:735)
	at java.base/jdk.internal.loader.URLClassPath$3.run(URLClassPath.java:499)
	at java.base/jdk.internal.loader.URLClassPath$3.run(URLClassPath.java:482)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:481)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:449)
	at java.base/jdk.internal.loader.URLClassPath$1.next(URLClassPath.java:345)
	at java.base/jdk.internal.loader.URLClassPath$1.hasMoreElements(URLClassPath.java:356)
	at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasNext(BuiltinClassLoader.java:408)
	at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasMoreElements(BuiltinClassLoader.java:416)
	at java.base/java.lang.CompoundEnumeration.next(ClassLoader.java:2725)
	at java.base/java.lang.CompoundEnumeration.hasMoreElements(ClassLoader.java:2734)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1210)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1228)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at java.base/sun.util.locale.provider.SPILocaleProviderAdapter$1.run(SPILocaleProviderAdapter.java:83)
	at java.base/sun.util.locale.provider.SPILocaleProviderAdapter$1.run(SPILocaleProviderAdapter.java:76)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at java.base/sun.util.locale.provider.SPILocaleProviderAdapter.findInstalledProvider(SPILocaleProviderAdapter.java:76)
	at java.base/sun.util.locale.provider.AuxLocaleProviderAdapter.getLocaleServiceProvider(AuxLocaleProviderAdapter.java:73)
	at java.base/sun.util.locale.provider.LocaleServiceProviderPool.findProviders(LocaleServiceProviderPool.java:304)
	at java.base/sun.util.locale.provider.LocaleServiceProviderPool.getLocalizedObjectImpl(LocaleServiceProviderPool.java:274)
	at java.base/sun.util.locale.provider.LocaleServiceProviderPool.getLocalizedObject(LocaleServiceProviderPool.java:256)
	at java.base/sun.util.locale.provider.CalendarDataUtility.retrieveFirstDayOfWeek(CalendarDataUtility.java:76)
	at java.base/java.util.Calendar.setWeekCountData(Calendar.java:3397)
	at java.base/java.util.Calendar.<init>(Calendar.java:1607)
	at java.base/java.util.GregorianCalendar.<init>(GregorianCalendar.java:738)
	at java.base/java.util.Calendar$Builder.build(Calendar.java:1492)
	at java.base/sun.security.util.DisabledAlgorithmConstraints$DenyAfterConstraint.<init>(DisabledAlgorithmConstraints.java:716)
	at java.base/sun.security.util.DisabledAlgorithmConstraints$Constraints.<init>(DisabledAlgorithmConstraints.java:428)
	at java.base/sun.security.util.DisabledAlgorithmConstraints.<init>(DisabledAlgorithmConstraints.java:148)
	at java.base/sun.security.util.DisabledAlgorithmConstraints.<init>(DisabledAlgorithmConstraints.java:122)
	at java.base/sun.security.util.DisabledAlgorithmConstraints$JarHolder.<clinit>(DisabledAlgorithmConstraints.java:98)
	... 84 more
```

**Comment by veita** [comment]
Yes, see above.

**Comment by veita** [comment]
Unfortunately not. But I'm rather sure that `-Djava.locale.providers=COMPAT,SPI` is in effect.

**Comment by veita** [comment]
A colleague of mine verified that removing `SPI` from `-Djava.locale.providers=COMPAT,SPI` in our case avoids the code path that leads to the `NullPointerException`.

However, the root cause of the error is the design flaw that there are cases in which initializing the class `sun.security.pkcs.SignerInfo` requires using `sun.security.pkcs.SignerInfo` itself, which is doomed to failure.

**Comment by veita** [comment]
This is the upstream bug report for this issue: https://bugs.openjdk.org/browse/JDK-8297684

If you like I would volunteer to backport the fix to 17.0.x.

**Comment by jerboaa** [comment]
https://bugs.openjdk.org/browse/JDK-8280890 is the bug that would need a backport to [jdk17u-dev](https://github.com/openjdk/jdk17u-dev). Feel free to propose it. I can help you get it in.

**Comment by veita** [comment]
Which repository should I fork such that a later pull request would be possible?

My guess would be https://github.com/adoptium/jdk17u.git, and `git checkout jdk-17.0.5+8_adopt`.

**Comment by jerboaa** [comment]
https://github.com/openjdk/jdk17u-dev and you'll have to have an OCA filed. We try to avoid downstream patches.

**Comment by jerboaa** [comment]
Finally have a reproducer: https://github.com/jerboaa/jdk/tree/sha1-disabled-spi-calendar-repro

**Comment by veita** [comment]
Filed pull request https://github.com/openjdk/jdk17u-dev/pull/947

**Comment by karianna** [comment]
Fixed in 17.0.6
